File: generate_geojson.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import io
import geojson
from geojson import Feature, Point, FeatureCollection
from generate_stationlist import STATION_FILENAME, load_station_list


STATION_ID_MAPPING_FILENAME = 'data/stations_id_mapping.csv'


def load_station_id_mapping(filename):
    mapping = {}
    with io.open(filename, 'r', encoding='utf-8') as f:
        for line in f:
            pref_id, station_id, name = line.strip().split('\t')
            mapping[name] = "/".join([pref_id, station_id])

    return mapping


def convert_feature(station, old_station_ids):
    if station.lat == 'None' or station.lng == 'None':
        return None

    geometry = Point((float(station.lng), float(station.lat)))
    properties = dict(pref_id=station.pref_id, station_id=station.station_id,
                      name=station.name, address=station.address,
                      hours=station.hours, uri=station.uri,
                      old_station_id=old_station_ids.get(station.name, None))
    return Feature(geometry=geometry, properties=properties)


def main():
    stations = load_station_list(STATION_FILENAME)
    old_statioN_ids = load_station_id_mapping(STATION_ID_MAPPING_FILENAME)
    features = (convert_feature(s, old_statioN_ids) for s in stations)
    with io.open('data/stations.geojson', 'w', encoding='utf-8') as f:
        json = geojson.dumps(FeatureCollection(list(filter(bool, features))),
                             ensure_ascii=False)
        f.write(json)


if __name__ == '__main__':
    main()

File: src/scripts/generate_geojson.ts
import * as fs from 'fs';
import * as StationCSV from '../lib/station-csv.js';
import type { Station } from '../lib/types.js';

const STATION_FILENAME = 'data/stations.csv';
const OUTPUT_FILENAME = 'data/stations.geojson';

interface GeoJSONFeature {
  type: 'Feature';
  geometry: {
    type: 'Point';
    coordinates: [number, number];
  };
  properties: {
    prefId: string;
    stationId: string;
    name: string;
    address: string;
    tel: string;
    hours: string;
    uri: string;
    mapcode: string;
  };
}

interface GeoJSONFeatureCollection {
  type: 'FeatureCollection';
  features: GeoJSONFeature[];
}


function convertFeature(station: Station): GeoJSONFeature | null {
  if (station.lat === 'None' || station.lng === 'None') {
    return null;
  }

  const lat = parseFloat(station.lat);
  const lng = parseFloat(station.lng);

  if (isNaN(lat) || isNaN(lng)) {
    return null;
  }

  return {
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: [lng, lat]
    },
    properties: {
      prefId: station.prefId,
      stationId: station.stationId,
      name: station.name,
      address: station.address,
      tel: station.tel,
      hours: station.hours,
      uri: station.uri,
      mapcode: station.mapcode
    }
  };
}

function main(): void {
  const stations = StationCSV.load(STATION_FILENAME);
  const features = stations
    .map(convertFeature)
    .filter((feature): feature is GeoJSONFeature => feature !== null);

  const featureCollection: GeoJSONFeatureCollection = {
    type: 'FeatureCollection',
    features
  };

  const json = JSON.stringify(featureCollection);
  fs.writeFileSync(OUTPUT_FILENAME, json, 'utf-8');
}

if (require.main === module) {
  main();
}
